okhttp-dnsoverhttps/src/main/java/okhttp3/dnsoverhttps/DnsOverHttps.java

```diff
@@ -0,0 +1,142 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package okhttp3.dnsoverhttps;
+
+import java.io.IOException;
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+import java.util.List;
+import javax.annotation.Nullable;
+import okhttp3.CacheControl;
+import okhttp3.Dns;
+import okhttp3.HttpUrl;
+import okhttp3.MediaType;
+import okhttp3.OkHttpClient;
+import okhttp3.Protocol;
+import okhttp3.Request;
+import okhttp3.RequestBody;
+import okhttp3.Response;
+import okhttp3.internal.platform.Platform;
+import okio.ByteString;
+
+/**
+ * DNS over HTTPS implementation.
+ *
+ * Implementation of https://tools.ietf.org/html/draft-ietf-doh-dns-over-https-09
+ *
+ * <blockquote>A DNS API client encodes a single DNS query into an HTTP request
+ * using either the HTTP GET or POST method and the other requirements
+ * of this section.  The DNS API server defines the URI used by the
+ * request through the use of a URI Template.</blockquote>
+ */
+public class DnsOverHttps implements Dns {
+  public static final MediaType DNS_MESSAGE = MediaType.parse("application/dns-message");
+  public static final MediaType UDPWIREFORMAT = MediaType.parse("application/dns-udpwireformat");
+  private final OkHttpClient client;
+  private final HttpUrl url;
+  private final boolean includeIPv6;
+  private final boolean post;
+  private final MediaType contentType;
+
+  public DnsOverHttps(OkHttpClient client, HttpUrl url, @Nullable Dns bootstrapDns,
+      boolean includeIPv6, String method, MediaType contentType) {
+    this.client = bootstrapDns != null ? client.newBuilder().dns(bootstrapDns).build() : client;
+    this.url = url;
+    this.includeIPv6 = includeIPv6;
+    if (!method.equals("GET") && !method.equals("POST")) {
+      throw new UnsupportedOperationException("Only GET and POST Supported");
+    }
+    this.post = method.equals("POST");
+    this.contentType = contentType;
+  }
+
+  public HttpUrl getUrl() {
+    return url;
+  }
+
+  public boolean isPost() {
+    return post;
+  }
+
+  @Override public List<InetAddress> lookup(String hostname) throws UnknownHostException {
+    try {
+      ByteString query = DnsRecordCodec.encodeQuery(hostname, includeIPv6);
+
+      Request request = buildRequest(query);
+      Response response = executeRequest(request);
+
+      return readResponse(hostname, response);
+    } catch (UnknownHostException uhe) {
+      throw uhe;
+    } catch (Exception e) {
+      UnknownHostException unknownHostException = new UnknownHostException(hostname);
+      unknownHostException.initCause(e);
+      throw unknownHostException;
+    }
+  }
+
+  private Response executeRequest(Request request) throws IOException {
+    // cached request
+    if (!post && client.cache() != null) {
+      Request cacheRequest = request.newBuilder().cacheControl(CacheControl.FORCE_CACHE).build();
```

Comment:
Oooh, not sure about this. `FORCE_CACHE` returns 503 if the response isn’t already in the cache.

Reply:
That's ok it falls through below.  I just don't want it to hit the network just to get a short circuited - no change response.  I want to strongly prefer the cache, and just fallback to network once it's stale.  Ideally continuing to use old response if server is now failing?

Reply:
Ahhhh. In that case I think you just want this:

```
CacheControl cacheControl = new CacheControl.Builder()
      .maxStale(Integer.MAX_VALUE, TimeUnit.SECONDS)
      .build();
```

That’ll use the cache if there’s anything there (even super stale). But if there’s nothing in the cache your request will still go through on the network.